pkg/cache/cache.go

```diff
@@ -0,0 +1,251 @@
+}
+
+// NewResultData creates a new [Result] for a successful operation.
+func NewResultData[T any](data T, etag string) Result[T] {
+	return Result[T]{
+		data: data,
+		etag: etag,
+	}
+}
+
+// NewResultError creates a new [Result] when an error has happened.
+func NewResultError[T any](err error) Result[T] {
+	return Result[T]{
+		err: err,
+	}
+}
+
+// Result can be treated as a [Interface] if necessary.
+func (r Result[T]) Data() Result[T] {
+	return r
+}
+
+// Split the result into its data, etag and error components.
+func (r Result[T]) Split() (T, string, error) {
+	return r.data, r.etag, r.err
+}
+
+// Interface is a cache that performs an action that will be cached. It
+// also returns an "etag" identifier to version the cache, so that the
+// caller can know if they have the most recent version of the cache
+// (and can decide to cache some operation based on that).
+//
+// The [NewMerger] and [NewTransformer] automatically handle
+// that for you by checking if the etag is updated before calling the
+// merging or transforming function.
+type Interface[T any] interface {
+	// Returns the cached data, as well as an "etag" to identify the
+	// version of the cache, or an error if something happened.
+	Data() Result[T]
+}
+
+// T is the source type, V is the destination type.
+type merger[K comparable, T, V any] struct {
+	mergeFn      func(map[K]Result[T]) Result[V]
+	caches       map[K]Interface[T]
+	cacheResults map[K]Result[T]
+	result       Result[V]
+}
+
+// NewMerger creates a new merge cache, a cache that merges the result
+// of other caches. The function only gets called if any of the
+// dependency has changed.
+//
+// If any of the dependency returned an error before, or any of the
+// dependency returned an error this time, or if the mergeFn failed
+// before, then the function is reran.
+//
+// The caches and results are mapped by string so that associated data
+// can be retrieved.
+func NewMerger[K comparable, T, V any](mergeFn func(results map[K]Result[T]) Result[V], caches map[K]Interface[T]) Interface[V] {
+	return &merger[K, T, V]{
+		mergeFn: mergeFn,
+		caches:  caches,
+	}
+}
+
+func (c *merger[K, T, V]) prepareResults() map[K]Result[T] {
+	cacheResults := make(map[K]Result[T], len(c.caches))
+	for key, cache := range c.caches {
+		cacheResults[key] = cache.Data()
+	}
+	return cacheResults
+}
+
+// Rerun if:
+//   - The last run resulted in an error
+//   - Any of the dependency previously returned an error
+//   - Any of the dependency just returned an error
+//   - Any of the dependency's etag changed
+func (c *merger[K, T, V]) needsRunning(results map[K]Result[T]) bool {
+	if c.cacheResults == nil {
+		return true
+	}
+	if c.result.err != nil {
+		return true
+	}
+	if len(results) != len(c.cacheResults) {
+		panic(fmt.Errorf("invalid number of results: %v (expected %v)", len(results), len(c.cacheResults)))
+	}
+	for key, oldResult := range c.cacheResults {
+		newResult, ok := results[key]
+		if !ok {
```

Comment:
Is this possible given that cacheResults and results should have the same length?

Reply:
Yeah, the keys can have changed. Say we cached when you had ApiService1 and ApiService2, and next time someone asks for the openapi we have ApiService1 and Apiservice3?

Reply:
Okay maybe I'm misunderstanding the code. From what I understand, the cache keys are added when the cache is created and do not change while the cache values change and are updated as needed. If the map can be modified at will, wouldn't L142 have cases where it'd fail?

```
	if len(results) != len(c.cacheResults) {
		panic(fmt.Errorf("invalid number of results: %v (expected %v)", len(results), len(c.cacheResults)))
	}
```

Reply:
Yeah, you're right, the map isn't supposed to change (we re-create the merger if we want to change the map), this should panic if the key isn't present.